Add nrf21540_ek shield support

KAGA164 opened the pull request:
This PR adds the nRF21540 EK Board shield definition. It is a continuation of 
https://github.com/zephyrproject-rtos/zephyr/pull/31523
which was targeted to upstream Zephyr but due to lack of use cases there, it was rejected.

NordicBuilder commented:
You can find the documentation preview for this PR at this <a href='http://developer.nordicsemi.com/nRF_Connect_SDK_dev/doc/PR-5263
'>link</a>. It will be updated about 10 minutes after the documentation build succeeds.

*Note: This comment is automatically posted by the Documentation Publishing GitHub Action.*

hubertmis reviewed (approved):
Looks good! :+1:

mia-ko reviewed (changes requested):
Just a few comments/suggestions.

More important, we should discuss moving the (doc) content to a location where it can be built and properly reviewed. Perhaps to a folder named boards in ncs/nrf/doc.

We should also talk about where this content should be displayed, in what context. There is a user guide "Working with FEM" that is planned to come in the near future where this topic could go. Until that is in place, it could maybe be added to the [FEM support](http://developer.nordicsemi.com/nRF_Connect_SDK/doc/latest/nrf/ug_radio_fem.html) UG.
From the tech writer perspective, this would be preferred location.

An alternative location that I have heard of is creating a Boards/Supported boards section on the top level where this can be added. Not sure about this one, but I would really like to hear your thoughts on this @carlescufi and @jdstapes.

Also, is having this board documented in NCS meant as a temporary location where we are planning to move it to Zephyr as soon as possible? Or is it a permanent solution?

KAGA164 commented on boards/shields/nrf21540_ek/doc/index.rst:
```diff
@@ -0,0 +1,84 @@
++21 dBm TX output power and 13 dB RX gain ensure a superior link budget for
+up to 16x range extension.
+
+The nRF21540 complementary device has a 50Ω SMA input and 2x 50Ω SMA outputs.
```
Done

KAGA164 commented on boards/shields/nrf21540_ek/doc/index.rst:
```diff
@@ -0,0 +1,84 @@
+
+The nRF21540's gain control, antenna switching, and modes are controlled via
+GPIO or SPI or a combination of both, accessible through
+the Arduino Uno Rev3 compatible headers.
```
Done

KAGA164 commented on boards/shields/nrf21540_ek/doc/index.rst:
```diff
@@ -0,0 +1,84 @@
+=================================
+
++-----------------------+----------+-----------------+
+| Shield Connector Pin  | SIGNAL   | FEM Function    |
```
Done

KAGA164 commented:
> Just a few comments/suggestions.
> 
> More important, we should discuss moving the (doc) content to a location where it can be built and properly reviewed. Perhaps to a folder named boards in ncs/nrf/doc.
> 
> We should also talk about where this content should be displayed, in what context. There is a user guide "Working with FEM" that is planned to come in the near future where this topic could go. Until that is in place, it could maybe be added to the [FEM support](http://developer.nordicsemi.com/nRF_Connect_SDK/doc/latest/nrf/ug_radio_fem.html) UG.
> From the tech writer perspective, this would be preferred location.
> 
> An alternative location that I have heard of is creating a Boards/Supported boards section on the top level where this can be added. Not sure about this one, but I would really like to hear your thoughts on this @carlescufi and @jdstapes.
> 
> Also, is having this board documented in NCS meant as a temporary location where we are planning to move it to Zephyr as soon as possible? Or is it a permanent solution?

It might be moved to Zephyr in future, moving this depends on a use cases of nRF21540 in Zephyr at this moments BLE controller there uses Kconfig instead if DTS if it stars using DTS then we can move shield support to the Zephyr. @carlescufi @jdstapes do you have any comment about where we should put documentation ?

hubertmis commented:
Would it be possible to merge shield files soon and clarify documentation later, in following PR? This way we could prevent creating overlays for this shield in samples, like #5294 proposed.

greg-fer commented:
> Would it be possible to merge shield files soon and clarify documentation later, in following PR? This way we could prevent creating overlays for this shield in samples, like #5294 proposed.

As discussed yesterday in a Teams chat with me, Lukasz, and Hubert:

- There are already overlays implemented for Zigbee and Thread samples since February.
- The #5294 is following in the wake of Zigbee and Thread samples.
- To make this PR and #5294 consistent, Hubert proposed the following PR that removes overlays: https://github.com/KAGA164/fw-nrfconnect-nrf/pull/8 -- to be merged into this PR, although it needs proper review first, as the documentation changes there are incomplete.

Questions still stand about the relationship between this PR from Kamil and the user guide @peknis is working on.

peknis commented:
Not exactly working on the actual FEM user guide (at least yet), but planning to add the nRF21540 DK to the requirements tables of all samples supporting nRF53840.
If we also need to add the current FEM support chapter, I believe I would also need to add the overlay files to the sample folders. And how about the sample.yaml files, would I need to amend it as for example in https://github.com/nrfconnect/sdk-nrf/pull/5294/files#diff-d49928f22e2d93edf4184858d9bfc98ebed7820455930bfd8ec1f563e20fcd67?

KAGA164 commented:
I added commit which removes obsolete overlay files for the Direct Test Mode and the Radio Test sample and also added a @hubertmis commit which do the same for Thread and Zigbee samples

grochu reviewed (approved):
Looks good to me

tomchy commented on boards/shields/nrf21540_ek/doc/index.rst:
```diff
@@ -0,0 +1,84 @@
+The FEM can be configured through the pins available on the Arduino headers.
+
+The nRF21540's gain control, antenna switching, and modes are controlled via
+GPIO or SPI, or a combination of both.
```
If I am not mistaken, the shield will always enable the SPI. I think it is worth mentioning here, that the peripheral device will be occupied, even if the GPIO-only interface is used

KAGA164 replied:
Yes it enables SPI always but maybe I will remove Kconfig.defconfig file which enables SPI and let user to decide whatever he want to use SPI also. What do you think @hubertmis ?

hubertmis replied:
I think it would be better. SPI should be enabled depending on selected nRF21540 driver. If given driver does not use SPI interface, there is no point in enabling SPI.

KAGA164 replied:
Done

tomchy replied:
Works for me!
What about the device tree overlay and the `spi-if` property?
Isn't it that way that if you set `spi-if = <&nrf_radio_fem_spi>;`, the driver will try to load SPI during the initialization?

KAGA164 replied:
It doesn't load driver because default value for CONFIG_SPI is n so drivers won't be added into build tree. This only creates a DTS macros which can be used in your application

KAGA164 replied:
Please note that the nrf21540dk board definition also added this in DTS

hubertmis replied:
Not necessarily. It is possible to have physical connection between nRF5 and nRF21540 SPI port and select nRF21540 GPIO driver. In such scenario, devicetree shall describe both GPIO and SPI connection, but the driver shall put SPI pins (at least CSN) to a known safe state. And this driver shall not enable SPI subsystem in Zephyr.

hubertmis replied:
So it should work in the following way:

Range of available nRF21540 drivers depends on entries in devicetree (GPIO, SPI, both, mixed, ...).
Driver's Kconfig should manage its dependencies (SPI, GPIO, whatever).

There should be no direct connection between nRF21540 device tree and driver dependencies, like "if you have SPI device, enable SPI driver".

tomchy replied:
Thank you for the clarification!

tomchy commented on boards/shields/nrf21540_ek/doc/index.rst:
```diff
@@ -0,0 +1,84 @@
++21 dBm TX output power and 13 dB RX gain ensure a superior link budget for
+up to 16x range extension.
+
+The nRF21540 complementary device has a 50 Ω SMA input and 2x 50 Ω SMA outputs.
```
I'm not sure about the input/output naming here, since we are receiving RF signals through output ports :slightly_smiling_face:

KAGA164 replied:
What about ?
 The nRF21540 complementary device has a 50 Ω SMA transceiver interface and 2x 50 Ω SMA antenna interfaces.

KAGA164 replied:
Done

hubertmis commented on samples/zigbee/ncp/sample.yaml:
```diff
@@ -27,6 +27,6 @@ tests:
     platform_allow: nrf52840dk_nrf52840
     tags: ci_build
-    extra_args: DTC_OVERLAY_FILE=boards/nrf52840dk_nrf52840.overlay;../common/dts-nrf21540-fem.overlay
+    extra_args: SHIELD=nrf21540_ek
```
But it's added automatically, because its name matches the board name. Isn't it?

KAGA164 replied:
Those overlays will be added automatically now

tomchy replied:
You're right! Shield does not modify board name, so this is no longer needed :+1:

greg-fer commented on doc/nrf/includes/sample_fem_support.txt:
```diff
@@ -5,30 +5,17 @@
 See :ref:`ug_radio_fem` for more information.
 
-To add support for the nRF21540 FEM, add the provided :file:`dts-nrf21540-fem.overlay` devicetree overlay file when building.
-The file is located in the |fem_file_path| folder.
```
Given that `|fem_file_path|` is removed, it should be also removed from all applicable samples (e.g. Thread and Zigbee), as it is used to provide custom overlay source file path in each sample doc.

KAGA164 replied:
Done

mia-ko reviewed (changes requested):
1. Still would like to see the index.rst file and the image be moved to a location in nrf/doc/nrf folder, so the documentation builds.
2. I see a lot of references to commands in command line when building a sample. It was pointed out to me that we should make sure that the SES option (which is the primary option officially) is covered.

KAGA164 commented:
>     1. Still would like to see the index.rst file and the image be moved to a location in nrf/doc/nrf folder, so the documentation builds.
> 
>     2. I see a lot of references to commands in command line when building a sample. It was pointed out to me that we should make sure that the SES option (which is the primary option officially) is covered.

@mia-ko Moving it there as an orphant doc entry probably causes that a doc build fails. My proposition is to make a shield entry in doc/nrf/index.rst toctree and  put this doc into doc/nrf/shield folder. Is it acceptable ?

mia-ko commented:
> > ```
> > 1. Still would like to see the index.rst file and the image be moved to a location in nrf/doc/nrf folder, so the documentation builds.
> > 
> > 2. I see a lot of references to commands in command line when building a sample. It was pointed out to me that we should make sure that the SES option (which is the primary option officially) is covered.
> > ```
> 
> @mia-ko Moving it there as an orphant doc entry probably causes that a doc build fails. My proposition is to make a shield entry in doc/nrf/index.rst toctree and put this doc into doc/nrf/shield folder. Is it acceptable ?

That is a top level entry in NCS, and I am not sure that it really makes sense... I feel that it should be incorporated with https://developer.nordicsemi.com/nRF_Connect_SDK/doc/latest/nrf/ug_radio_fem.html. At least for the time being.

But, what I also meant is that for the moment, the rst file and the png image are placed in boards folder which is not treated as source for docs so it doesn't build anyway. I think the file itself should end up in doc/ folder, somewhere. Otherwise it will not be visible, like it is not visible now. Correct me if I am wrong here.

KAGA164 commented:
@mia-ko I moved doc under https://developer.nordicsemi.com/nRF_Connect_SDK/doc/latest/nrf/ug_radio_fem.html

peknis commented on doc/nrf/shield/nrf21540_ek.rst:
```diff
@@ -0,0 +1,86 @@
+References
+==========
+
+.. target-notes::
+
+.. _nRF21540 website:
+   https://www.nordicsemi.com/Products/Low-power-short-range-wireless/nRF21540
+
+
+.. _nRF21540 Product Specification and other documentation:
+   https://infocenter.nordicsemi.com/topic/struct_fem/struct/nrf21540.html
```
```suggestion
* `nRF21540 DK product page`_
* `nRF21540 Product Specification`_
* `nRF21540`_
```

KAGA164 replied:
Done

mia-ko reviewed (commented):
The nRF21540 EK should probably be added to this overview as well https://developer.nordicsemi.com/nRF_Connect_SDK/doc/latest/nrf/app_boards.html

KAGA164 commented on doc/nrf/shield/nrf21540_ek.rst:
```diff
@@ -0,0 +1,86 @@
+Overview
+========
+
+.. image:: /images/nrf21540_ek.png
```
I move part of a description under the Overview section

KAGA164 commented on doc/nrf/ug_radio_fem.rst:
```diff
@@ -191,3 +191,7 @@ Then there is no need to define a GPIO to control the PDN signal. The line ``pdn
+
+.. _ug_radio_fem_nrf21540_ek:
+
+.. include:: /shield/nrf21540_ek.rst
```
Ok. I copied this file content here

KAGA164 commented:
@mia-ko @peknis I fixed a doc content and I just put it into https://developer.nordicsemi.com/nRF_Connect_SDK/doc/latest/nrf/ug_radio_fem.html. I think that we can change a location in separated PR after we get a decision where we should put this doc. Is it acceptable now ?

czeslawmakarski commented on doc/nrf/ug_radio_fem.rst:
```diff
@@ -191,3 +191,7 @@ Then there is no need to define a GPIO to control the PDN signal. The line ``pdn
 
 Generally, if pin ``X`` is not used, the ``X-gpios = < .. >;`` property can be removed.
```
Regarding the quoted sentence:
FEM support is something which can be implemented in the protocols which are using it. This FEM support implementation also depends on the family of the SoC (nRF52 or nRF53).
Protocols which do support FEM on the nRF52 family now (now it is for development, with the upmerge which is currently finishing the readiness shall be production-grade) are 802.15.4 (meaning both Thread and Zigbee) and BLE. Combination of BLE and 802.15.4 means that the multiprotocol operation is also supported.
Support for nRF53 family is something which is being developed right now and hopefully shall be present in the final release of NCS 1.7.0 (though nRF21540-GPIO version only).

Now, on to the board -DK and shield -EK. nRF21540-DK is supported as well as nRF21540-EK. While EK being a shield (and belonging to the shield subsystem), it is worth mentioning that proper `west` arguments are needed and link to the shields subsys documentation.

Now, to the

peknis commented on doc/nrf/ug_radio_fem.rst:
```diff
@@ -191,3 +191,7 @@ Then there is no need to define a GPIO to control the PDN signal. The line ``pdn
 Generally, if pin ``X`` is not used, the ``X-gpios = < .. >;`` property can be removed.
 This applies to all properties with a ``-gpios`` suffix, for both nRF21540 and SKY66112-11.
+
+.. _ug_radio_fem_nrf21540_ek:
```
@KAGA164 this is something we don't still see in the doc. Have a section "Supported boards" mentioning DK and EK and then nRF21450 EK under that. Otherwise, the EK story comes suddenly without any explanation.

mia-ko replied:
Sure, I can add a commit for this :)

KAGA164 replied:
@peknis I squashed Mia's commit. I hope that doc is good enough now :)

peknis commented on doc/nrf/includes/sample_dtm_radio_test_fem.txt:
```diff
@@ -1,33 +1,14 @@
 
 .. note::
-   You must set ``NRF21540EK_FEM=y`` if you use the nRF21540 EK.
+   If you use the nRF21540 EK, append `nrf21540_ek` shield to your build command instructing build system to append the appropriate devicetree overlay file.
```
```suggestion
   If you use the nRF21540 EK, append ``nrf21540_ek`` shield to your build command instructing build system to append the appropriate devicetree overlay file.
```

KAGA164 replied:
Done

peknis commented on doc/nrf/includes/sample_fem_support.txt:
```diff
@@ -5,30 +5,17 @@
-When you set the ``DTC_OVERLAY_FILE`` variable, you must specify all overlay files that are needed for building.
-Check the :file:`boards` folder to see the additional overlay files.
+.. note::
+   If you use the nRF21540 EK, append `nrf21540_ek` shield to your build command instructing build system to append the appropriate devicetree overlay file.
```
```suggestion
   If you use the nRF21540 EK, append ``nrf21540_ek`` shield to your build command instructing build system to append the appropriate devicetree overlay file.
```

KAGA164 replied:
Done

peknis commented on doc/nrf/app_boards.rst:
```diff
@@ -95,6 +95,9 @@ The following boards are defined in the :file:`nrf/boards/arm/` folder.
 |                   |            | :ref:`thingy91_nrf52840 <ug_thingy91>`                   | ``thingy91_nrf52840``                 |
 +-------------------+------------+----------------------------------------------------------+---------------------------------------+
 
+The following shields are defined in the :file:`nrf/boards/shields` folder.
+
+* :ref:`ug_radio_fem_nrf21540_ek`
```
The :ref:`nRF2154 EK shield <ug_radio_fem_nrf21540_ek>` is defined in the :file:`nrf/boards/shields` folder.

KAGA164 replied:
Done

greg-fer reviewed (approved):
Reviewed the fem_support.txt file one final time.
